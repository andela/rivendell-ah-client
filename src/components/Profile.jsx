import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

/**
 * @param { object } profile to be renderd
 * @param { string } owner the logged in user
 * @param { bool } errors from loading user profile
 * @returns { string } html string
 */
const Profile = ({ profile, owner, errors }) => {
  const { firstName, lastName, bio, username, email, image } = profile;
  let biography = 'Bio';
  let bioText = bio;
  let updateButton = '';
  if (owner === username) {
    updateButton = (
      <Link id="update-profile-link" to={`/@${username}/edit`}>
        Edit your Profile
      </Link>
    );
  }
  if (owner !== username) {
    updateButton = '';
  }
  if (!bio && (owner === username)) {
    bioText = `You have not added you biograpphy.
    Edit your profile and add you biograhpy`;
  }
  if (!bio && (owner !== username)) {
    biography = '404 bio not found';
  }
  let imageEle =
    <img id="profile-image" src={image} alt={username} />;
  if (!image) {
    imageEle = (
      <i id="alternate-profile-image" className="large material-icons">
        account_circle
      </i>
    );
  }

  let displayError = '';
  if (owner !== username && errors.message) {
    const { message } = errors;
    let displayMessage = '';
    if (message === 'User not found') {
      displayMessage = 'The user you are looking for does not exist';
    }

    if (message === 'Network Error') {
      displayMessage =
      'The server is down at the moment. Please try again later.';
    }
    displayError = <p id="profile-error-message">{displayMessage}</p>;
  }

  return (
    <div className="content">
      <div id="profile-header">
        <div id="profile-image-div">
          {imageEle}
        </div>
        <h3 id="profile-name">
          {firstName}
          {' '}
          {lastName}
        </h3>
        {displayError}
        <p id="email-text">{email}</p>
      </div>
      <div className="profile-detail-container">
        <div className="profile-details">
          <p id="bio-field">{biography}</p>
          <p id="bio-text">{bioText}</p>
        </div>
        {updateButton}
      </div>
    </div>
  );
};

/* eslint-disable react/forbid-prop-types */
Profile.propTypes = {
  errors: PropTypes.object,
  profile: PropTypes.object,
  owner: PropTypes.string.isRequired,
};

Profile.defaultProps = {
  errors: {},
  profile: {},
};

export default Profile;
