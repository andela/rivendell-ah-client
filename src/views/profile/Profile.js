import React from 'react';
import { Switch, Route } from 'react-router-dom';
import articlesPages from '../articles';
import EditProfile from './EditProfile';
import follows from '../follows'

const { Article, UserArticles, EditArticle } = articlesPages;
const { Followers, Following } = follows;



const Profile= () => (
  <div>
    <h2>The Profile of a User</h2>
    <Switch>
      <Route exact path='/:username/:slug' component={Article} />
      <Route exact path='/:username/new-article' component={EditArticle} />
      <Route exact path='/:username/:slug/edit' component={EditArticle} />
      <Route exact path='/:username/edit' component={EditProfile} />
      <Route exact path='/:username/articles' component={UserArticles} />
      <Route exact path='/:username/following' component={Following} />
      <Route exact path='/:username/followers' component={Followers} />
    </Switch>
  </div>
);

export default Profile;