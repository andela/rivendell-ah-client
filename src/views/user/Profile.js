import React from 'react';

const Profile = () => (
  <div>
    <h2> A User's Profile </h2>
  </div>
);

export default Profile;
