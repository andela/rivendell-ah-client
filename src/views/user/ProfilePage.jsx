import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { getProfile } from '../../actions/profile';
import Profile from '../../components/Profile.jsx';

/**
 * profile page
 * @return {string} rendered profile page
 */
export class ProfilePage extends Component {
  /**
   * dispatch the getprofile action when page load
   * @returns { object } user profile detail
   */
  componentDidMount() {
    const { username } = this.props.match.params;
    const loggedInUser = this.props.userProfile.username;
    if (loggedInUser !== username) {
      this.props.getProfile(`/profiles/${username}`);
    }
  }

  /**
   * render the profile page component and
   * pass down properties to the profile component
   * @returns {string} html string
   */
  render() {
    const { username } = this.props.match.params;
    let { profile } = this.props;
    const loggedInUser = this.props.userProfile.username;
    if (loggedInUser === username) {
      const { userProfile } = this.props;
      profile = userProfile;
    }

    return (
      <div id="profile-page">
        <Profile
          profile={profile}
          username={username}
          owner={this.props.userProfile.username}
          errors={this.props.errors}
        />
      </div>
    );
  }
}

/* eslint-disable react/forbid-prop-types */
ProfilePage.propTypes = {
  match: PropTypes.object.isRequired,
  profile: PropTypes.object,
  userProfile: PropTypes.object,
  getProfile: PropTypes.func.isRequired,
  errors: PropTypes.object,
};

ProfilePage.defaultProps = {
  errors: {},
  profile: {},
  userProfile: {},
};

export const mapStateToProps = (state) => {
  const { isLoading, errors, userProfile, profile } = state.profile;
  return {
    isLoading, errors, userProfile, profile,
  };
};

export default connect(mapStateToProps, { getProfile })(ProfilePage);
