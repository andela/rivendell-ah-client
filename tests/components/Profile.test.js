import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { ProfilePage, mapStateToProps } from '../../src/views/user/ProfilePage.jsx';
import Profile from '../../src/components/Profile.jsx';
import configureStore from 'redux-mock-store';


const profile = {
  bio: null,
  email: "reader11@gmail.com",
  firstName: "myfirs",
  id: 18,
  image: null,
  lastName: "mylas",
  username: "reader11",
  verified: true,
};

const owner = 'reader11';
const history = {
  push: jest.fn(),
  location: '/',
  listen: jest.fn(),
};
const getProfile = jest.fn()

const appState = {
  auth: {
    user: {},
    token: 'tokenvariable'
  },
  profile: {
    isLoading: false,
    errors: {},
    profile: {},
    userProfile: {},
  }
}

describe('Profile', () => {
  describe('component', () => {
    it('should render', () => {
      const match = {
        params: {
          username: 'reader11'
        }
      }
      const wrapper = shallow(
      <Profile
        userProfile={profile}
        profile={profile}
        match={match}
        owner={owner}
      />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with an alternate materialize icon if no image is present', () => {
      const match = {
        params: {
          username: 'reader11'
        }
      }
      const wrapper = mount(
        <MemoryRouter>
            <Profile
              profile={profile}
              owner={owner}
              match={match}
            />
        </MemoryRouter>
      )
      const div = wrapper.find('div');
      const materialIconAlt = div.find('[id="alternate-profile-image"]');
      expect(materialIconAlt.length).toBe(1);
    });
  });

  describe('Page', () => {
    it('should render', () => {
      const match = {
        params: {
          username: 'reader11'
        }
      }
      const wrapper = shallow(
      <ProfilePage
        owner={owner}
        match={match}
        getProfile={getProfile}
        userProfile={profile}
        profile={profile}
      />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with an update link', () => {
      const match = {
        params: {
          username: 'reader11'
        }
      }
      const wrapper = mount(
        <MemoryRouter>
          <ProfilePage
            owner={owner}
            match={match}
            getProfile={getProfile}
            userProfile={profile}
          />
        </MemoryRouter>
      );
      const updateLink = wrapper.find('Link [id="update-profile-link"]');
      expect(updateLink.length).toBe(1);
      expect(updateLink.text()).toEqual('Edit your Profile');
    });

    it('should render without an update profile link', () => {
      const match = {
        params: {
          username: 'reader1'
        }
      }
      const wrapper = mount(
        <MemoryRouter>
          <ProfilePage
            owner={owner}
            match={match}
            getProfile={getProfile}
            userProfile={profile}
            profile={profile}
          />
        </MemoryRouter>
      );
      const updateLink = wrapper.find('Link [id="update-profile-link"]');
      expect(updateLink.length).toBe(1);
    });
    it('should render with an alternate profile image', () => {

      const wrapper = mount(
        <MemoryRouter>
          <Profile profile={profile} owner={profile.username} errors={{}}/>
        </MemoryRouter>
      );
      const altProfileImage = wrapper.find('#alternate-profile-image');
      expect(altProfileImage.text())
      .toBe('account_circle');
    });

    it('should render with a custom error message if user is not found', () => {
      const errors = {
        message: 'User not found',
      }
      const profile = {};
      const owner = 'somebody'
      const wrapper = mount(
        <MemoryRouter>
          <Profile profile={profile} owner={owner} errors={errors}/>
        </MemoryRouter>
      );
      const errorMessage = wrapper.find('#profile-error-message');
      expect(errorMessage.text())
      .toBe('The user you are looking for does not exist');
    });

    it('should render with a custom error message if there is network error', () => {
      const errors = {
        message: 'Network Error',
      }
      const profile = {};
      const owner = 'somebody'
      const wrapper = mount(
        <MemoryRouter>
          <Profile profile={profile} owner={owner} errors={errors}/>
        </MemoryRouter>
      );
      const errorMessage = wrapper.find('#profile-error-message');
      expect(errorMessage.text())
      .toBe('The server is down at the moment. Please try again later.');
    });
  });

  describe('mapStateToProps', () => {
    it('should map the state to props correctly', () => {
      const componentState = mapStateToProps(appState);
      const expectedState = {
        errors: {},
        userProfile: {},
        isLoading: false,
        profile: {},
      }
      expect(componentState).toEqual(expectedState);
    })
  })
});
