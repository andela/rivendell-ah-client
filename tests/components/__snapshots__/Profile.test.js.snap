// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Page should render 1`] = `
<div
  id="profile-page"
>
  <Profile
    errors={Object {}}
    owner="reader11"
    profile={
      Object {
        "bio": null,
        "email": "reader11@gmail.com",
        "firstName": "myfirs",
        "id": 18,
        "image": null,
        "lastName": "mylas",
        "username": "reader11",
        "verified": true,
      }
    }
    username="reader11"
  />
</div>
`;

exports[`Profile component should render 1`] = `
<div
  className="content"
>
  <div
    id="profile-header"
  >
    <div
      id="profile-image-div"
    >
      <i
        className="large material-icons"
        id="alternate-profile-image"
      >
        account_circle
      </i>
    </div>
    <h3
      id="profile-name"
    >
      myfirs
       
      mylas
    </h3>
    <p
      id="email-text"
    >
      reader11@gmail.com
    </p>
  </div>
  <div
    className="profile-detail-container"
  >
    <div
      className="profile-details"
    >
      <p
        id="bio-field"
      >
        Bio
      </p>
      <p
        id="bio-text"
      >
        You have not added you biograpphy.
    Edit your profile and add you biograhpy
      </p>
    </div>
    <Link
      id="update-profile-link"
      replace={false}
      to="/@reader11/edit"
    >
      Edit your Profile
    </Link>
  </div>
</div>
`;
